radix-engine/src/system/transaction/multithread_txn_processor.rs

```diff
@@ -0,0 +1,138 @@
+use crate::blueprints::transaction_processor::{
+    ResumeResult, TxnProcessorThread, MAX_TOTAL_BLOB_SIZE_PER_INVOCATION,
+};
+use crate::errors::{KernelError, RuntimeError};
+use crate::kernel::kernel_callback_api::KernelCallbackObject;
+use crate::system::actor::{Actor, FunctionActor};
+use crate::system::system::SystemService;
+use crate::system::system_callback::{System, SystemBasedKernelApi};
+use crate::system::system_modules::auth::AuthModule;
+use radix_common::constants::TRANSACTION_PROCESSOR_PACKAGE;
+use radix_common::prelude::{BlueprintId, GlobalAddressReservation};
+use radix_engine_interface::blueprints::transaction_processor::{
+    TRANSACTION_PROCESSOR_BLUEPRINT, TRANSACTION_PROCESSOR_RUN_IDENT,
+};
+use radix_rust::prelude::*;
+use radix_transactions::model::{ExecutableTransaction, InstructionV2};
+use sbor::prelude::ToString;
+
+/// Multi-thread transaction processor for executing multiple subintents
+pub struct MultiThreadedTxnProcessor {
+    pub threads: Vec<(TxnProcessorThread<InstructionV2>, Vec<usize>)>,
+}
+
+impl MultiThreadedTxnProcessor {
+    pub fn init<Y: SystemBasedKernelApi>(
+        executable: ExecutableTransaction,
+        global_address_reservations: Vec<GlobalAddressReservation>,
+        api: &mut Y,
+    ) -> Result<Self, RuntimeError> {
+        let mut txn_processors = vec![];
+
+        // Setup
+        let intents = executable.intents();
+        for (thread_id, intent) in intents.iter().enumerate() {
+            api.kernel_switch_stack(thread_id)?;
+
+            let mut system_service = SystemService::new(api);
+            let virtual_resources = intent
+                .auth_zone_init
+                .simulate_every_proof_under_resources
+                .clone();
+            let virtual_non_fungibles =
+                intent.auth_zone_init.initial_non_fungible_id_proofs.clone();
+            let auth_zone = AuthModule::create_auth_zone(
+                &mut system_service,
+                None,
+                virtual_resources,
+                virtual_non_fungibles,
+            )?;
+
+            api.kernel_set_call_frame_data(Actor::Function(FunctionActor {
```

Comment:
Hmm this feels weird to me - is this right that we're _replacing_ the root call frame with this?

In principle, this isn't a problem, but I think we need to discuss what the abstraction is around `Actor::Root` and how it interacts with other parts of the codebase (cf comments on https://github.com/radixdlt/radixdlt-scrypto/pull/1919 )